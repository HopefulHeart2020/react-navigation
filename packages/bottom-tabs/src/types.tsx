import * as React from 'react';
import {
  AccessibilityRole,
  AccessibilityStates,
  StyleProp,
  TextStyle,
  ViewStyle,
} from 'react-native';
import {
  NavigationHelpers,
  NavigationProp,
  ParamListBase,
  Descriptor,
  Route,
} from '@react-navigation/core';
import { TabNavigationState } from '@react-navigation/routers';

export type BottomTabNavigationEventMap = {
  /**
   * Event which fires on tapping on the tab in the tab bar.
   */
  tabPress: undefined;
  /**
   * Event which fires on long press on the tab in the tab bar.
   */
  tabLongPress: undefined;
};

export type Orientation = 'horizontal' | 'vertical';

export type LabelPosition = 'beside-icon' | 'below-icon';

export type BottomTabNavigationHelpers = NavigationHelpers<
  ParamListBase,
  BottomTabNavigationEventMap
>;

export type BottomTabNavigationProp<
  ParamList extends ParamListBase,
  RouteName extends keyof ParamList = string
> = NavigationProp<
  ParamList,
  RouteName,
  TabNavigationState,
  BottomTabNavigationOptions,
  BottomTabNavigationEventMap
> & {
  /**
   * Jump to an existing tab.
   *
   * @param name Name of the route for the tab.
   * @param [params] Params object for the route.
   */
  jumpTo<RouteName extends Extract<keyof ParamList, string>>(
    ...args: ParamList[RouteName] extends (undefined | any)
      ? [RouteName] | [RouteName, ParamList[RouteName]]
      : [RouteName, ParamList[RouteName]]
  ): void;
};

export type BottomTabNavigationOptions = {
  /**
   * Title text for the screen.
   */
  title?: string;

  /**
   * Title string of a tab displayed in the tab bar or React Element
   * or a function that given { focused: boolean, tintColor: string } returns a React.Node, to display in tab bar.
   * When undefined, scene title is used. To hide, see tabBarOptions.showLabel in the previous section.
   */
  tabBarLabel?:
    | React.ReactNode
    | ((props: {
        focused: boolean;
        tintColor: string;
        horizontal: boolean;
      }) => React.ReactNode);

  /**
   * React Element or a function that given { focused: boolean, tintColor: string } returns a React.Node, to display in the tab bar.
   */
  tabBarIcon?:
    | React.ReactNode
    | ((props: {
        focused: boolean;
        tintColor: string;
        horizontal: boolean;
      }) => React.ReactNode);

  /**
   * Accessibility label for the tab button. This is read by the screen reader when the user taps the tab.
   * It's recommended to set this if you don't have a label for the tab.
   */
  tabBarAccessibilityLabel?: string;

  /**
   * ID to locate this tab button in tests.
   */
  tabBarTestID?: string;

  /**
   * Boolean indicating whether the tab bar is visible when this screen is active.
   */
  tabBarVisible?: boolean;

  /**
   * Buttton component to render for the tab items instead of the default `TouchableWithoutFeedback`
   */
  tabBarButtonComponent?: React.ComponentType<any>;
};

export type BottomTabDescriptor = Descriptor<
  ParamListBase,
  string,
  TabNavigationState,
  BottomTabNavigationOptions
>;

export type BottomTabDescriptorMap = {
  [key: string]: BottomTabDescriptor;
};

export type BottomTabNavigationConfig = {
  /**
   * Whether the screens should render the first time they are accessed. Defaults to `true`.
   * Set it to `false` if you want to render all screens on initial render.
   */
  lazy?: boolean;
  /**
   * Custom tab bar component.
   */
  tabBarComponent?: React.ComponentType<BottomTabBarProps>;
  /**
   * Options for the tab bar which will be passed as props to the tab bar component.
   */
  tabBarOptions?: BottomTabBarOptions;
};

export type BottomTabBarOptions = {
  /**
   * Whether the tab bar gets hidden when the keyboard is shown.
   */
  keyboardHidesTabBar?: boolean;
  /**
   * Color for the icon and label in the active tab.
   */
  activeTintColor?: string;
  /**
   * Color for the icon and label in the inactive tabs.
   */
  inactiveTintColor?: string;
  /**
   * Background olor for the active tab.
   */
  activeBackgroundColor?: string;
  /**
   * background color for the inactive tabs.
   */
  inactiveBackgroundColor?: string;
  /**
   * Whether label font should scale to respect Text Size accessibility settings.
   */
  allowFontScaling?: boolean;
  /**
   * Whether the tab label should be visible. Defaults to `true`.
   */
  showLabel?: boolean;
  /**
   * Whether the tab icon should be visible. Defaults to `true`.
   */
  showIcon?: boolean;
  /**
   * Style object for the tab label.
   */
  labelStyle?: StyleProp<TextStyle>;
  /**
   * Style object for the tab container.
   */
  tabStyle?: StyleProp<ViewStyle>;
  /**
   * Whether the label is renderd below the icon or beside the icon.
   * When a function is passed, it receives the device orientation to render the label differently.
   * By default, in `vertical` orinetation, label is rendered below and in `horizontal` orientation, it's renderd beside.
   */
  labelPosition?:
    | LabelPosition
    | ((options: { deviceOrientation: Orientation }) => LabelPosition);
  /**
   * Whether the label position should adapt to the orientation.
   */
  adaptive?: boolean;
  /**
   * Style object for the tab bar container.
   */
  style?: StyleProp<ViewStyle>;
};

export type BottomTabBarProps = BottomTabBarOptions & {
  state: TabNavigationState;
  descriptors: BottomTabDescriptorMap;
  navigation: NavigationHelpers<ParamListBase>;
  onTabPress: (props: { route: Route<string> }) => void;
  onTabLongPress: (props: { route: Route<string> }) => void;
  getAccessibilityLabel: (props: {
    route: Route<string>;
  }) => string | undefined;
  getAccessibilityRole: (props: {
    route: Route<string>;
  }) => AccessibilityRole | undefined;
  getAccessibilityStates: (props: {
    route: Route<string>;
    focused: boolean;
  }) => AccessibilityStates[];
  getButtonComponent: (props: {
    route: Route<string>;
  }) => React.ComponentType<any> | undefined;
  getLabelText: (props: {
    route: Route<string>;
  }) =>
    | ((scene: {
        focused: boolean;
        tintColor: string;
        orientation: 'horizontal' | 'vertical';
      }) => React.ReactNode | undefined)
    | React.ReactNode;
  getTestID: (props: { route: Route<string> }) => string | undefined;
  renderIcon: (props: {
    route: Route<string>;
    focused: boolean;
    tintColor: string;
    horizontal: boolean;
  }) => React.ReactNode;
  activeTintColor: string;
  inactiveTintColor: string;
};
