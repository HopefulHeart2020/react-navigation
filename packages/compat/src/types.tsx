import type {
  ParamListBase,
  NavigationProp,
  Route,
} from '@react-navigation/native';
import type * as helpers from './helpers';

export type CompatNavigationProp<
  NavigationPropType extends NavigationProp<ParamListBase>,
  ParamList extends ParamListBase = NavigationPropType extends NavigationProp<
    infer P,
    any,
    any,
    any,
    any
  >
    ? P
    : ParamListBase,
  RouteName extends Extract<keyof ParamList, string> = Extract<
    NavigationPropType extends NavigationProp<any, infer R> ? R : string,
    string
  >
> = Omit<NavigationPropType, keyof typeof helpers> &
  {
    [method in Extract<keyof NavigationPropType, keyof typeof helpers>]: (
      ...args: Parameters<typeof helpers[method]>
    ) => void;
  } & {
    state: Route<RouteName> & {
      routeName: RouteName;
    };
    getParam<T extends keyof ParamList[RouteName]>(
      paramName: T,
      defaultValue?: ParamList[RouteName][T]
    ): ParamList[RouteName][T];
    isFirstRouteInParent(): boolean;
    dangerouslyGetParent<
      T = NavigationProp<ParamListBase> | undefined
    >(): T extends NavigationProp<ParamListBase>
      ? CompatNavigationProp<T>
      : undefined;
  };

export type CompatNavigationOptions<
  NavigationPropType extends NavigationProp<ParamListBase>,
  ScreenOptions extends {} = NavigationPropType extends NavigationProp<
    any,
    any,
    any,
    infer O
  >
    ? O
    : {}
> =
  | ((options: {
      navigation: CompatNavigationProp<NavigationPropType>;
      navigationOptions: Partial<ScreenOptions>;
      screenProps: unknown;
    }) => ScreenOptions)
  | ScreenOptions;

export type CompatScreenType<
  NavigationPropType extends NavigationProp<ParamListBase>
> = React.ComponentType<{
  navigation: CompatNavigationProp<NavigationPropType>;
  screenProps: unknown;
}> & {
  navigationOptions?: CompatNavigationOptions<NavigationPropType>;
};

export type CompatRouteConfig<
  NavigationPropType extends NavigationProp<ParamListBase>,
  ParamList extends ParamListBase = NavigationPropType extends NavigationProp<
    infer P,
    any,
    any,
    any,
    any
  >
    ? P
    : ParamListBase
> = {
  [RouteName in keyof ParamList]:
    | React.ComponentType<any>
    | ((
        | { screen: React.ComponentType<any> }
        | { getScreen(): React.ComponentType<any> }
      ) & {
        navigationOptions?: CompatNavigationOptions<NavigationPropType>;
        params?: ParamList[RouteName];
      });
};
