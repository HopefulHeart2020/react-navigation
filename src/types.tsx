import * as BaseActions from './BaseActions';

export type CommonAction = BaseActions.Action;

export type NavigationState = {
  /**
   * Unique key for the navigation state.
   */
  key: string;
  /**
   * Index of the currently focused route.
   */
  index: number;
  /**
   * List if valid route names.
   */
  names: string[];
  /**
   * List of rendered routes.
   */
  routes: Array<Route & { state?: NavigationState }>;
};

export type InitialState = Omit<Omit<NavigationState, 'names'>, 'key'> & {
  key?: undefined;
  names?: undefined;
  state?: InitialState;
};

export type Route<RouteName = string> = {
  /**
   * Unique key for the route.
   */
  key: string;
  /**
   * User-provided name for the route.
   */
  name: RouteName;
  /**
   * Params for the route.
   */
  params?: object | void;
};

export type NavigationAction = {
  type: string;
};

export type Router<Action extends NavigationAction = NavigationAction> = {
  /**
   * Initialize full navigation state with a given partial state.
   */
  getInitialState(options: {
    screens: { [key: string]: ScreenProps };
    partialState?: NavigationState | InitialState;
    initialRouteName?: string;
  }): NavigationState;

  /**
   * Take the current state and action, and return a new state.
   * If the action cannot be handled, return `null`.
   */
  getStateForAction(
    state: NavigationState,
    action: Action | CommonAction
  ): NavigationState | null;

  getStateForChildUpdate(
    state: NavigationState,
    payload: {
      update: NavigationState;
      focus?: boolean;
    }
  ): NavigationState;

  /**
   * Whether the action should also change focus in parent navigator
   */
  shouldActionChangeFocus(action: Action | CommonAction): boolean;

  /**
   * Action creators for the router.
   */
  actionCreators: { [key: string]: (...args: any) => Action };
};

export type ParamListBase = { [key: string]: object | void };

class PrivateValueStore<T> {
  /**
   * TypeScript requires a type to be actually used to be able to infer it.
   * This is a hacky way of storing type in a property without surfacing it in intellisense.
   */
  // @ts-ignore
  private __private_value_type?: T;
}

export type NavigationHelpers<
  ParamList extends ParamListBase = ParamListBase
> = {
  /**
   * Dispatch an action to the router.
   */
  dispatch(action: NavigationAction): void;

  /**
   * Navigate to a route in current navigation tree.
   *
   * @param name Name of the route to navigate to.
   * @param [params] Params object for the route.
   */
  navigate<RouteName extends keyof ParamList>(
    ...args: ParamList[RouteName] extends void
      ? [RouteName]
      : [RouteName, ParamList[RouteName]]
  ): void;

  /**
   * Reset the navigation state to the provided state.
   * If a key is provided, the state with matching key will be reset.
   */
  reset(state: InitialState & { key?: string }): void;

  /**
   * Go back to the previous route in history.
   */
  goBack(): void;
} & PrivateValueStore<ParamList>;

export type NavigationProp<
  ParamList extends ParamListBase,
  RouteName extends keyof ParamList
> = NavigationHelpers<ParamList> & {
  /**
   * State for the child navigator.
   */
  state: Route<RouteName> &
    (ParamList[RouteName] extends void
      ? never
      : { params: ParamList[RouteName] });
};

export type CompositeNavigationProp<
  A extends NavigationHelpers<ParamListBase>,
  B extends NavigationHelpers<ParamListBase>
> = Omit<A & B, keyof NavigationHelpers<any>> &
  NavigationHelpers<
    (A extends NavigationHelpers<infer T> ? T : never) &
      (B extends NavigationHelpers<infer U> ? U : never)
  >;

export type Descriptor = {
  /**
   * Render the component associated with this route.
   */
  render(): React.ReactNode;

  /**
   * Options for the route.
   */
  options: Options;
};

export type Options = {
  /**
   * Title text for the screen.
   */
  title?: string;

  [key: string]: any;
};

export type ScreenProps<
  ParamList extends ParamListBase = ParamListBase,
  RouteName extends keyof ParamList = string
> = {
  /**
   * Route name of this screen.
   */
  name: RouteName;

  /**
   * Navigator options for this screen.
   */
  options?: Options;

  /**
   * Initial params object for the route.
   */
  initialParams?: ParamList[RouteName];
} & (
  | {
      /**
       * React component to render for this screen.
       */
      component: React.ComponentType<any>;
    }
  | {
      /**
       * Render callback to render content of this screen.
       */
      children: (props: any) => React.ReactNode;
    });

export type TypedNavigator<
  ParamList extends ParamListBase,
  Navigator extends React.ComponentType<any>
> = {
  Navigator: React.ComponentType<
    React.ComponentProps<Navigator> & {
      /**
       * Route to focus on initial render.
       */
      initialRouteName?: keyof ParamList;
    }
  >;
  Screen: React.ComponentType<ScreenProps<ParamList, keyof ParamList>>;
};
