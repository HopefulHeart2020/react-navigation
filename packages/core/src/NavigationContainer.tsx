import * as React from 'react';
import * as CommonActions from './CommonActions';
import EnsureSingleNavigator from './EnsureSingleNavigator';
import NavigationBuilderContext from './NavigationBuilderContext';
import ResetRootContext from './ResetRootContext';
import useFocusedListeners from './useFocusedListeners';
import useDevTools from './useDevTools';
import useStateGetters from './useStateGetters';

import {
  Route,
  NavigationState,
  InitialState,
  PartialState,
  NavigationAction,
  NavigationContainerRef,
  NavigationContainerProps,
} from './types';

type State = NavigationState | PartialState<NavigationState> | undefined;

const MISSING_CONTEXT_ERROR =
  "We couldn't find a navigation context. Have you wrapped your app with 'NavigationContainer'?";

export const NavigationStateContext = React.createContext<{
  isDefault?: true;
  state?: NavigationState | PartialState<NavigationState>;
  getState: () => NavigationState | PartialState<NavigationState> | undefined;
  setState: (
    state: NavigationState | PartialState<NavigationState> | undefined
  ) => void;
  key?: string;
  performTransaction: (action: () => void) => void;
}>({
  isDefault: true,

  get getState(): any {
    throw new Error(MISSING_CONTEXT_ERROR);
  },
  get setState(): any {
    throw new Error(MISSING_CONTEXT_ERROR);
  },
  get performTransaction(): any {
    throw new Error(MISSING_CONTEXT_ERROR);
  },
});

/**
 * Remove `key` and `routeNames` from the state objects recursively to get partial state.
 *
 * @param state Initial state object.
 */
const getPartialState = (
  state: InitialState | undefined
): PartialState<NavigationState> | undefined => {
  if (state === undefined) {
    return;
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const { key, routeNames, ...partialState } = state;

  // @ts-ignore
  return {
    ...partialState,
    stale: true,
    routes: state.routes.map(route => {
      if (route.state === undefined) {
        return route as Route<string> & {
          state?: PartialState<NavigationState>;
        };
      }

      return { ...route, state: getPartialState(route.state) };
    }),
  };
};

/**
 * Container component which holds the navigation state.
 * This should be rendered at the root wrapping the whole app.
 *
 * @param props.initialState Initial state object for the navigation tree.
 * @param props.onStateChange Callback which is called with the latest navigation state when it changes.
 * @param props.children Child elements to render the content.
 * @param props.ref Ref object which refers to the navigation object containing helper methods.
 */
const Container = React.forwardRef(function NavigationContainer(
  {
    initialState,
    onStateChange,
    independent,
    children,
  }: NavigationContainerProps,
  ref: React.Ref<NavigationContainerRef>
) {
  const parent = React.useContext(NavigationStateContext);

  if (!parent.isDefault && !independent) {
    throw new Error(
      "Looks like you have nested a 'NavigationContainer' inside another. Normally you need only one container at the root of the app, so this was probably an error. If this was intentional, pass 'independent={true}' explicitely."
    );
  }

  const [state, setNavigationState] = React.useState<State>(() =>
    getPartialState(initialState == null ? undefined : initialState)
  );

  const navigationStateRef = React.useRef<State>();
  const transactionStateRef = React.useRef<State | null>(null);
  const isTransactionActiveRef = React.useRef<boolean>(false);
  const isFirstMountRef = React.useRef<boolean>(true);
  const skipTrackingRef = React.useRef<boolean>(false);

  const reset = React.useCallback((state: NavigationState) => {
    skipTrackingRef.current = true;
    setNavigationState(state);
  }, []);

  const { trackState, trackAction } = useDevTools({
    name: '@react-navigation',
    reset,
    state,
  });

  const { listeners, addListener: addFocusedListener } = useFocusedListeners();

  const { getStateForRoute, addStateGetter } = useStateGetters();

  const dispatch = (
    action: NavigationAction | ((state: NavigationState) => NavigationAction)
  ) => {
    listeners[0](navigation => navigation.dispatch(action));
  };

  const canGoBack = () => {
    const { result, handled } = listeners[0](navigation =>
      navigation.canGoBack()
    );

    if (handled) {
      return result;
    } else {
      return false;
    }
  };

  const resetRoot = React.useCallback(
    (state?: PartialState<NavigationState> | NavigationState) => {
      trackAction('@@RESET_ROOT');
      setNavigationState(state);
    },
    [trackAction]
  );

  const getRootState = React.useCallback(() => {
    return getStateForRoute('root');
  }, [getStateForRoute]);

  React.useImperativeHandle(ref, () => ({
    ...(Object.keys(CommonActions) as Array<keyof typeof CommonActions>).reduce<
      any
    >((acc, name) => {
      acc[name] = (...args: any[]) =>
        dispatch(
          CommonActions[name](
            // @ts-ignore
            ...args
          )
        );
      return acc;
    }, {}),
    resetRoot,
    dispatch,
    canGoBack,
    getRootState,
  }));

  const builderContext = React.useMemo(
    () => ({
      addFocusedListener,
      addStateGetter,
      trackAction,
    }),
    [addFocusedListener, trackAction, addStateGetter]
  );

  const performTransaction = React.useCallback((callback: () => void) => {
    if (isTransactionActiveRef.current) {
      throw new Error(
        "Only one transaction can be active at a time. Did you accidentally nest 'performTransaction'?"
      );
    }

    setNavigationState((navigationState: State) => {
      isTransactionActiveRef.current = true;
      transactionStateRef.current = navigationState;

      callback();

      isTransactionActiveRef.current = false;

      return transactionStateRef.current;
    });
  }, []);

  const getState = React.useCallback(
    () =>
      transactionStateRef.current !== null
        ? transactionStateRef.current
        : navigationStateRef.current,
    []
  );

  const setState = React.useCallback((navigationState: State) => {
    if (transactionStateRef.current === null) {
      throw new Error(
        "Any 'setState' calls need to be done inside 'performTransaction'"
      );
    }

    transactionStateRef.current = navigationState;
  }, []);

  const context = React.useMemo(
    () => ({
      state,
      performTransaction,
      getState,
      setState,
    }),
    [getState, performTransaction, setState, state]
  );

  React.useEffect(() => {
    if (skipTrackingRef.current) {
      skipTrackingRef.current = false;
    } else {
      trackState(getRootState);
    }

    navigationStateRef.current = state;
    transactionStateRef.current = null;

    if (!isFirstMountRef.current && onStateChange) {
      onStateChange(getRootState());
    }

    isFirstMountRef.current = false;
  }, [state, onStateChange, trackState, getRootState]);

  return (
    <NavigationBuilderContext.Provider value={builderContext}>
      <NavigationStateContext.Provider value={context}>
        <ResetRootContext.Provider value={resetRoot}>
          <EnsureSingleNavigator>{children}</EnsureSingleNavigator>
        </ResetRootContext.Provider>
      </NavigationStateContext.Provider>
    </NavigationBuilderContext.Provider>
  );
});

export default Container;
