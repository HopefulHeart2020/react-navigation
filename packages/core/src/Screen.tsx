import { ParamListBase } from '@react-navigation/routers';
import { RouteConfig } from './types';

/**
 * Empty component used for specifying route configuration.
 */
export default function Screen<
  ParamList extends ParamListBase,
  RouteName extends keyof ParamList,
  ScreenOptions extends object
>(_: RouteConfig<ParamList, RouteName, ScreenOptions>) {
  /* istanbul ignore next */
  return null;
}
