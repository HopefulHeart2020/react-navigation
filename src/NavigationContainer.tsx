import * as React from 'react';
import * as BaseActions from './BaseActions';
import EnsureSingleNavigator from './EnsureSingleNavigator';
import NavigationBuilderContext from './NavigationBuilderContext';
import useChildActionListeners from './useChildActionListeners';
import {
  Route,
  NavigationState,
  InitialState,
  PartialState,
  ParamListBase,
  NavigationHelpers,
  NavigationAction,
} from './types';

type Props = {
  initialState?: InitialState;
  onStateChange?: (
    state: NavigationState | PartialState<NavigationState> | undefined
  ) => void;
  children: React.ReactNode;
};

type State = NavigationState | PartialState<NavigationState> | undefined;

const MISSING_CONTEXT_ERROR =
  "We couldn't find a navigation context. Have you wrapped your app with 'NavigationContainer'?";

export const NavigationStateContext = React.createContext<{
  state?: NavigationState | PartialState<NavigationState>;
  getState: () => NavigationState | PartialState<NavigationState> | undefined;
  setState: (state: NavigationState | undefined) => void;
  key?: string;
  performTransaction: (action: () => void) => void;
}>({
  get getState(): any {
    throw new Error(MISSING_CONTEXT_ERROR);
  },
  get setState(): any {
    throw new Error(MISSING_CONTEXT_ERROR);
  },
  get performTransaction(): any {
    throw new Error(MISSING_CONTEXT_ERROR);
  },
});

const getPartialState = (
  state: InitialState | undefined
): PartialState<NavigationState> | undefined => {
  if (state === undefined) {
    return;
  }

  // @ts-ignore
  return {
    ...state,
    stale: true,
    key: undefined,
    routeNames: undefined,
    routes: state.routes.map(route => {
      if (route.state === undefined) {
        return route as Route<string> & {
          state?: PartialState<NavigationState>;
        };
      }

      return { ...route, state: getPartialState(route.state) };
    }),
  };
};

/**
 * Container component which holds the navigation state.
 * This should be rendered at the root wrapping the whole app.
 *
 * @param props.initialState Initial state object for the navigation tree.
 * @param props.onStateChange Callback which is called with the latest navigation state when it changes.
 * @param props.children Child elements to render the content.
 * @param props.ref Ref object which refers to the navigation object containing helper methods.
 */
const Container = React.forwardRef(function NavigationContainer(
  { initialState, onStateChange, children }: Props,
  ref: React.Ref<NavigationHelpers<ParamListBase>>
) {
  const [state, setNavigationState] = React.useState<State>(() =>
    getPartialState(initialState)
  );

  const dispatch = (action: NavigationAction) =>
    context.performTransaction(() => {
      for (let i = 0; i < actionListeners.length; i++) {
        actionListeners[i](action, undefined, null);
      }
    });

  React.useImperativeHandle(ref, () => ({
    dispatch,
    ...(Object.keys(BaseActions) as Array<keyof typeof BaseActions>).reduce<
      any
    >((acc, name) => {
      acc[name] = (...args: any[]) =>
        dispatch(
          // eslint-disable-next-line import/namespace
          BaseActions[name](
            // @ts-ignore
            ...args
          )
        );
      return acc;
    }, {}),
  }));

  const navigationStateRef = React.useRef<State>();
  const transactionStateRef = React.useRef<State | null>(null);
  const isTransactionActiveRef = React.useRef<boolean>(false);
  const isFirstMountRef = React.useRef<boolean>(true);

  const {
    listeners: actionListeners,
    addActionListener,
    removeActionListener,
  } = useChildActionListeners();

  const navigationContext = React.useMemo(
    () => ({
      addActionListener,
      removeActionListener,
    }),
    [addActionListener, removeActionListener]
  );

  const performTransaction = React.useCallback((callback: () => void) => {
    if (isTransactionActiveRef.current) {
      throw new Error(
        "Only one transaction can be active at a time. Did you accidentally nest 'performTransaction'?"
      );
    }

    setNavigationState((navigationState: State) => {
      isTransactionActiveRef.current = true;
      transactionStateRef.current = navigationState;

      callback();

      isTransactionActiveRef.current = false;

      return transactionStateRef.current;
    });
  }, []);

  const getState = React.useCallback(
    () =>
      transactionStateRef.current !== null
        ? transactionStateRef.current
        : navigationStateRef.current,
    []
  );

  const setState = React.useCallback((navigationState: State) => {
    if (transactionStateRef.current === null) {
      throw new Error(
        "Any 'setState' calls need to be done inside 'performTransaction'"
      );
    }

    transactionStateRef.current = navigationState;
  }, []);

  const context = React.useMemo(
    () => ({
      state,
      performTransaction,
      getState,
      setState,
    }),
    [getState, performTransaction, setState, state]
  );

  React.useEffect(() => {
    navigationStateRef.current = state;
    transactionStateRef.current = null;

    if (!isFirstMountRef.current && onStateChange) {
      onStateChange(state);
    }

    isFirstMountRef.current = false;
  }, [state, onStateChange]);

  return (
    <NavigationStateContext.Provider value={context}>
      <NavigationBuilderContext.Provider value={navigationContext}>
        <EnsureSingleNavigator>{children}</EnsureSingleNavigator>
      </NavigationBuilderContext.Provider>
    </NavigationStateContext.Provider>
  );
});

export default Container;
