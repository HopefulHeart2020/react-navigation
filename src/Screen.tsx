import { ScreenProps } from './types';

export default function Screen(_: ScreenProps) {
  return null;
}
