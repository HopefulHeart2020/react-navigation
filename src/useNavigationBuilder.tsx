import * as React from 'react';
import { NavigationStateContext } from './NavigationContainer';
import Screen from './Screen';
import useRegisterNavigator from './useRegisterNavigator';
import useDescriptors from './useDescriptors';
import useNavigationHelpers from './useNavigationHelpers';
import useOnAction from './useOnAction';
import { Router, NavigationState, ScreenProps } from './types';
import useOnChildUpdate from './useOnChildUpdate';
import useChildActionListeners from './useChildActionListeners';

type Options = {
  initialRouteName?: string;
  children: React.ReactNode;
};

const isArrayEqual = (a: any[], b: any[]) =>
  a.length === b.length && a.every((it, index) => it === b[index]);

export default function useNavigationBuilder(
  router: Router<any>,
  options: Options
) {
  useRegisterNavigator();

  const screens = React.Children.map(options.children, child => {
    if (child === null || child === undefined) {
      return;
    }

    if (React.isValidElement(child) && child.type === Screen) {
      return child.props as ScreenProps;
    }

    throw new Error(
      `A navigator can only contain 'Screen' components as its direct children (found '${
        // @ts-ignore
        child.type && child.type.name ? child.type.name : String(child)
      }')`
    );
  })
    .filter(Boolean)
    .reduce(
      (acc, curr) => {
        acc[curr!.name] = curr as ScreenProps;
        return acc;
      },
      {} as { [key: string]: ScreenProps }
    );

  const routeNames = Object.keys(screens);
  const initialRouteName =
    options.initialRouteName !== undefined
      ? options.initialRouteName
      : routeNames[0];
  const initialParamsList = routeNames.reduce(
    (acc, curr) => {
      acc[curr] = screens[curr].initialParams;
      return acc;
    },
    {} as { [key: string]: object | undefined }
  );

  const [initialState] = React.useState(() =>
    router.getInitialState({
      routeNames,
      initialRouteName,
      initialParamsList,
    })
  );

  const {
    state: currentState = initialState,
    getState: getCurrentState,
    setState,
    key,
  } = React.useContext(NavigationStateContext);

  let state = router.getRehydratedState({
    routeNames,
    partialState: currentState,
  });

  if (!isArrayEqual(state.routeNames, routeNames)) {
    // When the list of route names change, the router should handle it to remove invalid routes
    const nextState = router.getStateForRouteNamesChange(state, {
      routeNames,
      initialRouteName,
      initialParamsList,
    });

    if (state !== nextState) {
      // If the state needs to be updated, we'll schedule an update with React
      // setState in render seems hacky, but that's how React docs implement getDerivedPropsFromState
      // https://reactjs.org/docs/hooks-faq.html#how-do-i-implement-getderivedstatefromprops
      setState(nextState);
    }

    state = nextState;
  }

  React.useEffect(() => {
    return () => {
      // We need to clean up state for this navigator on unmount
      getCurrentState() !== undefined && setState(undefined);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const getState = React.useCallback(
    (): NavigationState =>
      router.getRehydratedState({
        routeNames,
        partialState: getCurrentState() || state,
      }),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [getCurrentState, router.getRehydratedState, router.getInitialState]
  );

  const {
    listeners: actionListeners,
    addActionListener,
    removeActionListener,
  } = useChildActionListeners();

  const onAction = useOnAction({
    router,
    getState,
    setState,
    key,
    getStateForAction: router.getStateForAction,
    actionListeners,
  });

  const onChildUpdate = useOnChildUpdate({
    router,
    onAction,
    key,
    getState,
    setState,
  });

  const helpers = useNavigationHelpers({
    onAction,
    getState,
    setState,
    actionCreators: router.actionCreators,
  });

  const navigation = React.useMemo(() => ({ ...helpers, state }), [
    helpers,
    state,
  ]);

  const descriptors = useDescriptors({
    state,
    screens,
    helpers,
    onAction,
    getState,
    setState,
    onChildUpdate,
    addActionListener,
    removeActionListener,
  });

  return {
    navigation,
    descriptors,
  };
}
