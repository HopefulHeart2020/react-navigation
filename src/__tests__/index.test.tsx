import * as React from 'react';
import { render } from 'react-native-testing-library';
import NavigationContainer from '../NavigationContainer';
import useNavigationBuilder from '../useNavigationBuilder';
import { Router } from '../types';
import Screen from '../Screen';

const MockRouter: Router = {
  getInitialState({ screens, initialRouteName }) {
    const routeNames = Object.keys(screens);

    return {
      key: 'root',
      names: routeNames,
      index: routeNames.indexOf(initialRouteName || routeNames[0]),
      routes: routeNames.map(name => ({
        key: name,
        name,
      })),
    };
  },

  getStateForAction() {
    return null;
  },

  getStateForChildUpdate(state) {
    return state;
  },

  shouldActionPropagateToChildren() {
    return false;
  },

  shouldActionChangeFocus() {
    return false;
  },

  actionCreators: {},
};

it('initializes state with router', () => {
  expect.assertions(1);

  const TestNavigator = (props: any) => {
    useNavigationBuilder(MockRouter, props);
    return null;
  };

  const element = (
    <NavigationContainer
      onStateChange={state => expect(state).toMatchSnapshot()}
    >
      <TestNavigator initialRouteName="bar">
        <Screen name="foo" component={jest.fn()} />
        <Screen name="bar" component={jest.fn()} />
        <Screen name="baz" component={jest.fn()} />
      </TestNavigator>
    </NavigationContainer>
  );

  render(element).update(element);
});

it('throws if muliple navigators rendered under one container', async () => {
  expect.assertions(1);

  const TestNavigator = (props: any) => {
    useNavigationBuilder(MockRouter, props);
    return null;
  };

  const element = (
    <NavigationContainer>
      <TestNavigator>
        <Screen name="foo" component={jest.fn()} />
      </TestNavigator>
      <TestNavigator>
        <Screen name="foo" component={jest.fn()} />
      </TestNavigator>
    </NavigationContainer>
  );

  expect(() => render(element).update(element)).toThrowError(
    'Another navigator is already registered for this container'
  );
});
