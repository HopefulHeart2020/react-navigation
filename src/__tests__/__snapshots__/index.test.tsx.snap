// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initializes state with router 1`] = `
Object {
  "index": 1,
  "key": "root",
  "names": Array [
    "foo",
    "bar",
    "baz",
  ],
  "routes": Array [
    Object {
      "key": "foo",
      "name": "foo",
    },
    Object {
      "key": "bar",
      "name": "bar",
    },
    Object {
      "key": "baz",
      "name": "baz",
    },
  ],
}
`;
